mx.py

```diff
@@ -4223,51 +4239,89 @@ def deploy_binary(args):
-    parser.add_argument('repository', action='store', help='Repository name used for Maven deploy (must be defined in a suite.py file)')
+    parser.add_argument('--all-suites', action='store_true', help='Perform binary deploy also for dependent suites')
+    parser.add_argument('--skip-existing', action='store_true', help='Do not deploy if already in repository')
+    parser.add_argument('repository_id', metavar='repository-id', action='store', help='Repository ID used for binary deploy')
+    parser.add_argument('url', metavar='repository-url', nargs='?', action='store', help='Repository URL used for binary deploy, if no url is given, the repository-id is looked up in suite.py')
     args = parser.parse_args(args)
 
-    s = _primary_suite
-    if not s.getMxCompatibility().supportsLicenses():
-        log("Not deploying '{0}' because licenses aren't defined".format(s.name))
+    suites = OrderedDict()
+
+    def import_visitor(s, suite_import, **extra_args):
+        suite_collector(suite(suite_import.name), suite_import)
+
+    def suite_collector(s, suite_import):
+        if s in suites:
+            return
+        suites[s] = None
+        s.visit_imports(import_visitor)
+
+    primary_suite = _check_primary_suite()
+    if args.all_suites:
+        suite_collector(primary_suite, None)
+    else:
+        suites[primary_suite] = None
+
+    for s in iter(suites):
+        _deploy_binary(args, s)
+
+def _deploy_binary(args, suite):
+    if not suite.getMxCompatibility().supportsLicenses():
+        log("Not deploying '{0}' because licenses aren't defined".format(suite.name))
         return
-    if not s.getMxCompatibility().supportsRepositories():
-        log("Not deploying '{0}' because repositories aren't defined".format(s.name))
+    if not suite.getMxCompatibility().supportsRepositories():
+        log("Not deploying '{0}' because repositories aren't defined".format(suite.name))
         return
-    if not s.vc:
-        abort('Current prinary suite has no version control')
+    if not suite.vc:
+        abort('Current primary suite has no version control')
     _mvn.check()
     def _versionGetter(suite):
         return '{0}-SNAPSHOT'.format(suite.vc.parent(suite.dir))
-    dists = s.dists
+    dists = suite.dists
     if args.only:
         only = args.only.split(',')
         dists = [d for d in dists if d.name in only]
     if args.platform_dependent:
         dists = [d for d in dists if d.platformDependent]
 
-    mxMetaName = _mx_binary_distribution_root(s.name)
-    s.create_mx_binary_distribution_jar()
-    mxMetaJar = s.mx_binary_distribution_jar_path()
+    mxMetaName = _mx_binary_distribution_root(suite.name)
+    suite.create_mx_binary_distribution_jar()
+    mxMetaJar = suite.mx_binary_distribution_jar_path()
     assert exists(mxMetaJar)
+    if args.all_suites:
+        dists = [d for d in dists if d.exists()]
```

Comment:
Well, but then shouldn't we check that instead of checking if the distribution exists?
Otherwise this disables the safety guard that all the necessary distributions will be uploaded.

Also, this means the actual semantics of `--all-suites` is "this suites and the distributions it depends on in other suites".